Add Agentic Coding Tile (#25)
Ada-Developers-Academy/independent-ai-topics: agentic-coding/best-practices.md

Reviewer: While I was following along and generating some chat output (e.g., in order to see the Restore Checkpoint button), I ran into rate limiting in OpenRouter. I first ran into it using Nvidia Nemotron 3 Ultra (free), though it sounds like the entire free tier has a 20 request per minute limit.

It doesn't sound like there's a way to tell copilot to rate limit itself, but with some agentic help, I was able to put together a proxy that injected a 3s delay per request (guaranteeing no more than 20 requests per minute).

Long story short, I'm wondering whether we should bring up potential rate limiting issues and how they could be worked around (the proxy was non-trivial, but maybe there's a more canned approach?). Maybe just reiterate here that folks should primarily use the core Copilot-provided models and that the OpenRouter setup performed earlier is more of a fallback and that they might not have as smooth an experience with it?

Author: I added a suggestion for an info callout to the top of the set up steps lesson, does this content address what you're hoping?

> ## Prefer Copilot over OpenRouter
> 
> Copilot integrates better with VS Code and is what we recommend folks use first when trying out agentic coding. Credits are limited for Copilot access through the GitHub Student Developer Pack, so we have included OpenRouter set up as a fallback option in case we run out of Copilot credits. 
> 
> OpenRouter has rate limiting for free tiers that can cause issues with VS Code when those limits are hit and the agent cannot finish a request in progress. Typically we can recover by rolling back to the most recent checkpoint if we run into those kinds of problems.

Reviewer: The verbiage sounds good, and the placement in the setup probably makes the most sense. This lesson is the one where I starterd actively working through examples, but there were earlier lessons where this could have already been an issue.

I'm still wondering whether a brief reminder at the start of the subsequent lessons might be helpful, but it's hard to say what the right balance would be.

# Best Practices

We have some foundations under our belt now: what agents are and how they work, how context windows determine the output and usefulness of a session, and the shape of workflows that experienced practitioners tend to converge on. 

This lesson shifts focus from *what* to do toward *how* to do it well. Best practices don't exist independently of the systems they support, so we'll connect the practices described here back to the workflow we've already seen.

## Learning Goals

> - Identify controls for adusting the course of an existing agent session

Reviewer: ```suggestion
- Identify controls for adjusting the course of an existing agent session
```

- Distinguish what belongs in a steering file versus a skill, and explain how that distinction affects context window usage.
- Define the structure of a skill and write a basic skill file.
- Explain how model selection affects workflow performance and cost.
- Identify when subagents are worth the added complexity and describe how to define one.

## Vocabulary and Synonyms

| Vocab | Definition | Synonyms | How to Use in a Sentence |
| --------- | --------- | -------- | --------- |

> | YAML frontmatter | A structured block of metadata written in YAML syntax that appears at the top of a markdown file, enclosed between two lines of triple dashes (`---`). | Front matter, metadata block | "The YAML frontmatter in our skill file sets the name and description the agent uses to decide when to load it." |

Reviewer: Minor annoyance, but the spacing of the frontmatter description results in the triple dash being line-wrapped. We could either rephrase the definition a bit, try to set styling on the column (maybe?), or in this suggestion, use the html `code` rather than markdown tick and explicitly make it no-wrap.

```suggestion
| YAML frontmatter | A structured block of metadata written in YAML syntax that appears at the top of a markdown file, enclosed between two lines of triple dashes (<code style="white-space: nowrap;">---</code>). | Front matter, metadata block | "The YAML frontmatter in our skill file sets the name and description the agent uses to decide when to load it." |
```

## Course Correcting: Checkpoints and Influencing Execution

Sometimes we might see that our agent has veered off-path, or we forgot something in the original prompt. We don't have to wait until the agent is done processing to take action; especially if we can see that the work will not meet our needs, we can act immediately!

### Checkpoints

Every response an agent produces in a chat session can serve as a checkpoint for our workspace. If we want to undo an agent's changes, we can restore an earlier checkpoint, which returns our files to the state they were in at that moment and removes any conversation history and file changes that came after it. To do so, we can hover over a message in the chat, and "Restore Checkpoint" controls will appear.

These checkpoints give us a way to explore a direction with an agent and fully back out of it if the result isn't what we needed, rather than manually tracking and reversing each change ourselves.

![VS Code chat window showing the "Restore Checkpoint" button after an agent response](assets/best-practices/vscode_chat_checkpoint.png)   

> *Fig. "Restore Checkpoint" button in the VS Code Chat UI ([Full Size Image](assets/best-practices/vscode_chat_checkpoint.png))*

Reviewer: I see that the next image down the page doesn't have the Full Size Image link. For me, I'd probably have left it off for both of these, since in neither one is there a lot of small text that we need the reader to review. This one is primarily showing where a UI element is.

It's not a problem to have the link here, but I'm wondering on what you tend to base including the link, for some level setting.

Author: I don't have a strict rubric, but whether I include a full size image link is based on how big the image is, and how much text there is (larger image or more text = add link). The main point of this image isn't most of the text, and I do think it would be fine without the full size link. 

As someone who looks at our curriculum on phones, tablets, and smaller screens pretty regularly, and who has issues related to my migraines that make me need to increase text or image sizes some days to get my eyes to focus on them properly, I can also see someone opening this larger just to be able to see the UI and context the image is sharing better. 

I tend to lean towards adding a link if it's on the edge, because I don't think it adds much overhead or inconvenience for us, but it can really help out folks who might need it.

Reviewer: Oh. That mobile and accessibility perspective is really helpful. I'll try to keep that in mind myself.

### !callout-info

## Try it out!

The next time we're working with an agent and want to explore a direction we might not keep, try backing out of it with a checkpoint instead of undoing the changes ourselves.

1. Send a prompt to our agent and let it produce at least one full response.
2. Hover over that response in the chat history and look for the "Restore Checkpoint" control.
3. Restore the checkpoint and confirm our files and conversation return to the state they were in before that response.

### !end-callout

### Steering During Execution

> While an agent is working, the send button in the chat pane becomes a drop down with options for how our message should be handled:

Reviewer: I'm on VS Code 1.131.0 and I don't get the drop down options unless there is additional text written into the chat box. If the box is empty, I just get the cancel button. I wasn't sure whether that was reflected in the current learn screen shot (is it showing a capital I or an input bar?), but it might be worth mentioning that this feature appears only after starting to type a new message?

<img width="408" height="138" alt="Image" src="https://github.com/user-attachments/assets/7344e283-5333-49ed-a771-325f3654e66a" />

Author: I was a version behind you, but I'm pretty sure something needs to be entered. I updated my VS Code to 1.131.0 & only see the drop down option when something is entered, so how does the following suggestion sound?
```suggestion
If we start to type a new message while an agent is working, the send button in the chat pane becomes a drop down with options for how our message should be handled:
```

Reviewer: Sounds good to me.

- "**Add to queue**" - Waits for the current response to finish, then send the message. Best used when the AI is mostly on track but we want to add something we forgot without stopping or changing the current progress.
- "**Steer with message**" - Tells the agent to pause after its current tool call and process our message immediately before continuing. This is useful when we want to adjust behavior without cancelling work that has already been done. 

> - "**Stop and send**" - Cancels the current request, removing work done for this request that is not presisted to disk. This is the right option when we need to start over, typically when the AI is far off track and continuing would waste resources.

Reviewer: ```suggestion
- "**Stop and send**" - Cancels the current request, removing work done for this request that is not persisted to disk. This is the right option when we need to start over, typically when the AI is far off track and continuing would waste resources.
```

![VS Code chat window showing the dropdown of send options while an agent is creating a response](assets/best-practices/vscode_chat_send_options.png)   
*Fig. VS Code Chat UI showing the Send options while an agent is creating a response*

### !callout-info

## Try it out!

The next time we're running an agent session, try each of the send options from the dropdown so we can feel the difference between them.

1. Start a task and, while the agent is still working, use "Add to queue" to send a small addition without interrupting its current progress.
2. On a separate task, use "Steer with message" to redirect the agent partway through once we notice it heading somewhere we don't want.
3. If we see an agent go far enough off track that the work in progress isn't worth keeping, try "Stop and send" to cancel and restart with a corrected prompt.

Compare how each option affects the agent's progress and note which situations call for which option.

### !end-callout

## Adding Instructions: Steering Files and Skills

> Earlier we introduced steering files and skills at a high level, here, we'll look at how to use them effectively. Both steering files and skills are ways of providing agents with context and instructions. Understanding the distinction between them and when to add to a steering file or create a new skill matters a great deal for context window usage and overall agent performance.

Reviewer: ```suggestion
Earlier we introduced steering files and skills at a high level. Here, we'll look at how to use them effectively. Both steering files and skills are ways of providing agents with context and instructions. Understanding the distinction between them and when to add to a steering file or create a new skill matters a great deal for context window usage and overall agent performance.
```

Strong steering file candidates:
- Rules and prohibitions that apply to everything: branch naming, files that must never be modified, required tests before committing
- Project or team-specific conventions the model wouldn't know from training, like naming patterns or import conventions
- Pointers to canonical examples in the codebase the model should reference when producing similar work
- Project structure and folder organization 
    - Depending on the size of the mapping/our code base, this architectural information can be offloaded to it's own file if every agent does not need to know the structure of the code base at all times.

What often steers people wrong is treating the steering file like a knowledge base. Team members start adding edge case documentation, onboarding notes, guides for specific workflows, and the file grows to several thousand tokens. Because it's loaded on every message for the entire session, this overhead compounds continuously. 

> - A bloated steering file is one of the most consistently expensive things we can do to our token usage.

Reviewer: I'm personally not a fan of lone bullet points. For things that are like a mini-callout, I like using blockquotes. This ends up emphasizing the material without being a "list" of one.

Of course, an actual callout would be fine as well, but when the emphasis is short enough to not need  a heading (sure would be nice if Learn supported "bare" callouts), the blockquote does a pretty nice job.

\*edit: There are a few more single bullter spots, as well as a blockquote being used as a prompt input a bit later on. I've added suggestions for them as well.

```suggestion
> A bloated steering file is one of the most consistently expensive things we can do to our token usage.
```

Author: I'm a little bit torn. I think the block quote formatting is really useful when it's a piece of actionable information because it feels like we're visually putting some emphasis on it. I like the idea of using that formatting for sentences like the one you highlighted here and the sentence later on: "Scripts should be treated with the same security awareness as any other executable code: they run in the agent's execution environment and belong inside a sandbox.". 

The extra highlighting I perceive when using block quotes feels weird to me for less actionable sentences like "Everything we corrected becomes a step or a gotcha in the skill file." or "As we get more experience with how our workflow runs, we can refine our agent choices further!"

How would you feel about leaving the less actionable sentences as they are now?

Reviewer: I'd rather have them all go in one direction or the other.

If a text unit feels to emphasized with the blockquote, does it really need the bullet? To me, the options would be

1. blockquote of actionable advice, new non-bullet paragraph for non-actionable (since a new paragraph is its own light emphasis).
2. just leave everything as it is. Like I said, it feels odd to me, but we only have so many tools with markdown. 🤷

Author: Since it would require editing all 6 lesson files to be consistent within the topic and it isn't a trivial find & replace change, from those options, I'm inclined to keep the formatting as-is through this lesson, but I will take that feedback into account for future content.

Steering files often live at the root of a project repo and are named `AGENTS.md`, `CLAUDE.md`, or similar depending on the environment we are working in. As we get more comfortable with steering contents, we can look into settings to apply steering files at the workspace or user level if there is steering information we find useful to apply to all projects we work with.

Let's take a look at an example steering file, the one below is for a TypeScript project:

```markdown
# Project Conventions

This is a Node.js/TypeScript project using Express for routing and PostgreSQL via pg-promise. `package.json` contains the approved packages in use for this project.

## Structure
- `src/routes/` — API route handlers
- `src/services/` — Business logic layer
- `src/db/` — Query files and migration scripts
- `src/generated/` — Auto-generated types. Do not modify directly.

## Standards
- All public functions must have JSDoc comments
- Run `npm test` to execute the test suite and ensure all tests pass before committing
- Use the service layer for business logic; route handlers should stay thin
- Branch naming: `feat/<ticket>`, `fix/<ticket>`, `chore/<description>`
```

> This file is short enough that it adds minimal overhead per message, but helps orient an agent picking up any task in the project so they can work within the team's norms.

Reviewer: matches singular "an agent"
```suggestion
This file is short enough that it adds minimal overhead per message, but helps orient an agent picking up any task in the project so it can work within the team's norms.
```

- Unlike the steering file, skills load a light index at session start up and the full contents of skills are only loaded when an agent encounters a task that matches the skill's description. They allow a large library of team knowledge to be available to agents without deeply impacting token usage until that knowledge is needed.

- Skills are shareable across projects and agents. A code review skill, a deployment checklist skill, or an API documentation skill developed once can be reused across every project that benefits from it.

#### Skill Structure

Skill files may live in different locations depending on our IDE. 
1. Typically there will be a hidden folder in the root of a project named something like `.agents`, `.claude`, or `.cursor` where we can create a folder named `skills` to add our custom skills.
2. For each skill, we create a new folder named after the skill and create a file named `SKILL.md` in that folder
3. Each `SKILL.md` file must contain a name, a short description of the skill it provides, and the steps the AI must take to complete the described task.

The basic structure of a skill file looks like:

> ```markdown

Reviewer: I think the markdown formatting looks a bit off in Learn due to the yaml header. But it looks pretty good if we tell it the file is yaml, even with the following markdown.

```suggestion
```yaml
```

---
name: database-migration
description: Use when creating a new Alembic migration, modifying the database 
             schema, or troubleshooting migration conflicts.
---

## Steps

1. Create the migration with `alembic revision --autogenerate -m "<description>"`.
2. Review the generated file in `migrations/versions/`. Autogenerate is not always accurate.
3. Check that `upgrade()` and `downgrade()` are both implemented correctly.
4. Test locally with `alembic upgrade head` before committing.
5. Note in the PR that a migration is included so reviewers know to run it.
```

> The name and description fields act as the trigger mechanism: the description needs to be specific enough that the agent can reliably identify when the skill applies (i.e. ("use when creating a new API endpoint")). 

Reviewer: Here, I was thinking it would be good to explicitly mention the naming convention, as well as how the name can be used, so I reworked it a bit.

```suggestion
The name and description fields act as triggers for the skill.
- A user can explicitly invoke a skill by its name (usually written in `kebab-case`) when prefaced with a `/`.
- An agent can invoke a skill by matching a task to a skill's description. The description therefore needs to be specific enough that the agent can reliably identify when the skill applies (e.g., "use when creating a new API endpoint").
```

> - A description that is too vague ("general coding help") will either never load or load when it doesn't apply. 

Reviewer: Follow on from line 88
```suggestion
> A description that is too vague ("general coding help") will either never load or load when it doesn't apply. 
```

> After the required name and description, we structure the steps for the agent to complete a task. Skill file sizes will range depending on what they describe, but in general, SKILL.md files should be focused on a single task and under roughly 500 lines.

Reviewer: ```suggestion
After the required name and description, we structure the steps for the agent to complete a task. Skill file sizes will range depending on what they describe, but in general, `SKILL.md` files should be focused on a single task and under roughly 500 lines.
```

##### Skills Can Include More Than Instructions

Beyond the main `SKILL.md` file, the skill folder structure supports:

- **References**: A `references/` subdirectory for supporting documentation like style guides, example input/output files, long configuration templates, and detailed references. These are also progressively disclosed, resources in the references folder are fetched on demand by the agent when it needs more depth on a step of a skill, they are not automatically loaded when the skill is used.
- **Scripts**: A `scripts/` subdirectory for executable code the agent can run as part of the skill workflow.

Scripts are particularly useful when a task involves a precise sequence of commands that needs to run the same way every time. Rather than relying on the agent to produce the command correctly from its training data, the script encodes exactly what runs. 

> - Scripts should be treated with the same security awareness as any other executable code: they run in the agent's execution environment and belong inside a sandbox.

Reviewer: Follow on from line 88
```suggestion
> Scripts should be treated with the same security awareness as any other executable code: they run in the agent's execution environment and belong inside a sandbox.
```

##### The Walkthrough Approach

The process starts with a fresh session and a real instance of the task. We give the agent the goal and watch what happens:
- Where does it make correct choices without instruction?
- Where does it go wrong or produce something inconsistent with what we want?
- What steps does it invent on its own that we don't actually want?
- Where does it get stuck or ask for clarification?

At each point where the agent deviates from what we want, we correct it in the session: "No, check the staging environment first before touching production configs" or "The output should be a markdown table, not a prose summary." We keep going until the agent completes the task in a way we'd be comfortable seeing repeated.

That corrected walkthrough becomes the skill! We've now observed the actual decision points that matter for this task in our specific environment, and we've identified which ones the agent won't get right on its own. 

> - Everything we corrected becomes a step or a gotcha in the skill file.

Reviewer: Follow on from line 88
```suggestion
> Everything we corrected becomes a step or a gotcha in the skill file.
```

**Ask the agent to document what it just did**

Once a walkthrough is complete and we're satisfied with the result, we can ask the agent to draft the skill file directly from the session: 

> > "Review what you just did to complete this task. Write a SKILL.md that would guide an agent through the same process reliably."

Reviewer: Follow on from line 88 (avoiding the blockquote since I was just using it for emphasis, whereas here we're providing a prompt)
````suggestion
```txt
Review what you just did to complete this task.
Write a SKILL.md that would guide an agent through
the same process reliably.
```
````

TIL that you can make a suggestion that includes a triple tick by using a quad tick for the suggestion block!

Author: I did not know that about the quad tics, and that makes life so much easier, thanks for sharing!

```

This kind of detail can be hard to derive from planning alone, it comes from running the task and observing where things break. The walkthrough approach builds those observations directly into the skill creation process rather than treating them as corrections to fix after the fact.

### !callout-info

## Try it out!

The "Token Usage & Context Windows" lesson covered end-of-session summaries as a way to preserve continuity between sessions. Rather than drafting a session-summary skill from memory, let's build it the way we just covered: through a walkthrough.

If you don't have a project with an active agent session open currently, open a project, start a session, and ask a few questions, so that there is some content in the chat to summarize.

1. At the end of an agent session, ask the agent to write a session summary to disk, using the categories from the "Token Usage & Context Windows" lesson as a starting point.
2. Review the result and correct anything missing or off-target directly in the session.
3. Once the summary reflects what we actually want, ask the agent to draft a `SKILL.md` from the session, using a prompt like the one covered above: 

>     > "Review what you just did to complete this task. Write a SKILL.md that would guide an agent through the same process reliably."

Reviewer: Follow on from line 88. Admittedly, this doesn't look great. Potentially, we could just omit requoting the prompt since it was just given above.
````suggestion
   ```txt
   Review what you just did to complete this task.
   Write a SKILL.md that would guide an agent through
   the same process reliably.
   ```
````

Author: I'd be good removing this one since it's so close in proximity to the initial statement of the prompt.

> 4. Find or create a skills directory in the current project, then create a `session-summary` folder in that directory. Save the draft created by the agent as `SKILL.md`, and clarify the name and description fields if needed.

Reviewer: ```suggestion
4. Find or create a `skills` directory in the current project, then create a `session-summary` folder in that directory. Save the draft created by the agent as `SKILL.md`, and clarify the name and description fields if needed.
```

Reviewer: Also, should we mention the top-level `.agents` folder?

Author: Good point, since the .agents folder hasn't really been mentioned since the start of the skills section, we probably should add something about creating the skills directory inside the top-level .agents folder.

> 4. Find or create a skills directory in the current project, then create a `session-summary` folder in that directory. Save the draft created by the agent as `SKILL.md`, and clarify the name and description fields if needed.

Author: Combines removing the repeated prompt & mentioning the top level .agents folder

```suggestion
3. Once the summary reflects what we actually want, ask the agent to draft a `SKILL.md` from the session, using a prompt like the one covered above.
4. Find or create a `skills` directory in the current project inside of the top-level `.agents` folder (folks may need to do some independent research on how to make hidden files visible). Inside the `skills` directory, create a `session-summary` folder in that directory. Save the draft created by the agent as `SKILL.md`, and clarify the name and description fields if needed.
```

A general framework for choosing models looks like:

| Task Type | Required Capabilities | Model Characteristics |
|-------|--------------|----------------------|
| Planning / architecture | Nuanced analysis, handling ambiguity, considering tradeoffs | Higher-capability models are worth the cost here; errors in the plan compound and are expensive to fix downstream |
| Implementation | Reliable code generation to spec | Mid-tier models are often sufficient for routine implementation given a given a clear and well-specified plan |
| Research / exploration | Reading and summarizing files, retrieving documentation | Smaller, faster models handle this well and cost significantly less |
| Boilerplate / scaffolding | Generating repetitive structures that follow a template | Smallest capable model; A lighter model that can follow an existing pattern is sufficient for this kind of work. |

Model choice doesn't require active management for every task. Many practitioners choose a default model for their primary session and a lighter model for subagents doing routine work. Some IDEs even have auto-model selection options that try to match appropriate models for a task based on our prompt. 

> - As we get more experience with how our workflow runs, we can refine our agent choices further!

Reviewer: Follow on from line 88.
```suggestion
> As we get more experience with how our workflow runs, we can refine our agent choices further!
```

As an example of how this can look in practice, let's assume we have a team that is building out a new notification service. They might map AI models to the tasks in the project like so: 
1. A higher-capability model is used to draft the architecture and implementation plan with a human reviewing the results at each step.
2. As part of planning, a research subagent uses a smaller model to pull documentation for the notification library and summarize existing related code in the project. 
3. Once the plan is finalized, the team has a subagent using a mid-tier model implement the code. 
4. The orchestrating session, still running the higher-capability model, coordinates results from subagents and keeps track of the progress through project tasks.

### Using Subagents

The reality is that as IDEs and agent harnesses continue to develop, they are getting more sophisticated around when to spin work off to subagents without us directly saying "Do some task in a subagent session". More and more frequently, using a subagent is not an explicit choice we are making, but is what's happening with our agentic coding tool set under the hood. 

> - If you have experimented with agentic coding before, it's possible that you've already experienced working with subagents in a way that abstracted away the subagent sessions!

Reviewer: Follow on from line 88.
```suggestion
> If you have experimented with agentic coding before, it's possible that you've already experienced working with subagents in a way that abstracted away the subagent sessions!
```

Most agentic coding tools ship with a default general-purpose agent. That default is a reasonable starting point, but as our workflows mature, we may find that certain tasks benefit from having specific configurations locked in at start up. 

A custom agent is a saved profile that combines a name, a model preference, a set of permitted tools, and a system prompt. We define the agent file once and activate it by name whenever the same kind of work comes up. This allows us to switch roles rather than rebuild configurations from scratch!

#### Agent Files

Custom agents are typically defined as markdown files. The file location varies by tool but is usually a project-level folder like `.github/agents/`, `.claude/agents/`, or a user-level directory for agents that should be available across all projects.
- Project-level agents live in the repository and are available to the whole team. 
- User-level agents are stored in a personal directory and follow us across projects. 

For team workflows, project-level is generally preferred: the agent configuration versions alongside the code, updates are shared automatically, and new team members get the configuration without setup.

> An agent definition has two parts: 

Reviewer: Callback to skills since they have similar structure.
```suggestion
Similar to skills, an agent definition has two parts: 
```

1. a YAML frontmatter block that sets configuration
2. a body that contains the system prompt

For example, a custom agent for code reviewing might look something like:

> ```markdown

Reviewer: Like in skills, I think this looks nicer with yaml formatting.
```suggestion
```yaml
```

Write your findings to `docs/reviews/<branch-name>-review.md` with a PASS or FAIL
verdict and an itemized list of issues.
```

The key fields:

- **`name`**: How we refer to and activate this agent
- **`description`**: What the agent is for; some tools surface this in the UI
- **`tools`**: The specific tools this agent is allowed to use. Restricting tools here is how we enforce phase-appropriate access without relying on the agent to self-limit.
- **`model`**: The model to use for this agent's sessions. This is where we can pin a higher-capability model for planning work and a lighter one for agents doing routine tasks.
- **`body`**: The prompt loaded on start up that defines the agent's behavior and any specific instructions. The body prompt should be specific enough that the agent behaves consistently without us restating context every time we activate it.

**When a custom agent is the right tool**

A one-off task rarely justifies a custom agent. The overhead of defining a custom agent pays off when a workflow recurs often enough that rebuilding its configuration manually each time becomes a friction point. 

> - When we find ourselves repeatedly selecting the same model, enabling the same tools, and restating the same framing before getting to work, that's the strongest signal that a custom agent would serve us better. 

Reviewer: Follow on from line 88.
```suggestion
> When we find ourselves repeatedly selecting the same model, enabling the same tools, and restating the same framing before getting to work, that's the strongest signal that a custom agent would serve us better. 
```

> Another signal is when the stakes of misconfiguration are high enough that we don't want to rely on getting it right each time. A planning agent that accidentally has write access, or a review agent that's running a smaller model than the task warrants, produces subtly worse results in ways that aren't always obvious in the moment. Encoding the right model and the right tool permissions into a named agent removes that variability. 

Reviewer: A little reflowing
```suggestion
Another signal is when the stakes of misconfiguration are high enough that we don't want to risk missing something each time we run the workflow. A planning agent that accidentally has write access, or a review agent that's running a smaller model than the task warrants, produces subtly worse results in ways that aren't always obvious in the moment. Encoding the right model and the right tool permissions into a named agent removes that variability. 
```

##### !end-question
##### !options

a| Manually track and reverse each file change we believe the agent made.
b| Hover over the message from the point we want to return to and select "Restore Checkpoint."
c| Close the session and start an entirely new one with a fresh prompt.
d| Wait for the agent to finish its current response before taking any action.

##### !end-options
##### !answer

b|

##### !end-answer
##### !explanation

Every response an agent produces serves as a checkpoint. Restoring a checkpoint returns our files to the state they were in at that point and removes any conversation history and file changes that came after it. This gives us a way to fully back out of a direction we explored, rather than manually tracking and reversing each change ourselves.

##### !end-explanation
### !end-challenge
<!-- prettier-ignore-end -->

<!-- prettier-ignore-start -->
### !challenge
* type: multiple-choice
* id: H7tY2sVbK9mQwL4xN0pR6cJ3zB
* title: Best Practices
##### !question

> While an agent is actively working through a task, we realize we forgot to include a constraint in our original prompt. The agent's progress so far still looks correct, and we don't want to stop or change the work already in progress. Which option should we choose from the send button's dropdown?

Reviewer: The phrasing here was a bit confusing for me. I interpretted the not stopping or chaning the work in progress as ruling out Stop and send, but talking about things being correct so far led me in the direction of we should update the session with the additional constraint before it gets much farther. So I went with Steer with message, since it waits until the current tool call is complete, but could head off the agent going down the wrong road if we wait until the full request is complete.

Maybe we could push more towards the desired Add to queue by mentioning that the work in progress looks like it's staying correct and is nearly done, so it's unlikely to misstep as it wraps up the request. Something like that?

Author: I fleshed out the scenario some to help motivate towards "Add to Queue"; how does that clarification land?

Reviewer: I think that helps. I was wondering whether we could take it a step further and say something like how we are aware that the omitted constraint currently only has instances it applies to in the third folder. Something along those lines that make it explicitly clear that there's no risk of letting the current run (now checkpinted to the first folder) complete. But phrasing that feels a little bulky to me, so I think this is fine.

##### !end-question
##### !options

a| Writing a longer skill description so the agent has more context about when to load it.
b| Running a live session where the actual deployment task is performed and correcting the agent's output at each misstep, then building the skill from that session.
c| Adding examples of the desired final outputs in a `references/` subdirectory.
d| Writing a second skill for just the step where the agent has issues.

##### !end-options
##### !answer

b|

##### !end-answer
##### !explanation

The walkthrough approach works through the task with a live agent session and corrects missteps as they happen. Those corrections become steps and gotchas in the skill file. Writing a skill from memory skips this discovery process, which means edge cases and environment-specific failure points, like the one this developer encountered, often don't get documented until they surface during actual use.

> 

Reviewer: Break to separate the paragraphs.
```suggestion

<br>

```
